### PR TITLE
Add All-Pairs Bellman-Ford

### PR DESCRIPTION
**Pull request by IvanIsCoding**
Follow up of #603 

This PR adds:
* `all_pairs_bellman_ford_path_lengths`
* `all_pairs_bellman_ford_shortest_paths`

They are the equivalent of the `all_pairs_dijkstra_path_lengths` and `all_pairs_dijkstra_shortest_paths` but for Bellman-Ford

### CODE REVIEW

File: src/shortest_path/all_pairs_bellman_ford.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                return (
+                    x.index(),
+                    PathLengthMapping {
+                        path_lengths: DictMap::new(),
+                    },
+                );
+            }
+
+            let out_map = PathLengthMapping {
+                path_lengths: path_lenghts
```
Comment (mtreinish) [inline src/shortest_path/all_pairs_bellman_ford.rs:87]:
```suggestion
                path_lengths: path_lengths
```

File: src/shortest_path/all_pairs_bellman_ford.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    let negative_cycle = RwLock::new(false);
+
+    let out_map = AllPairsPathMapping {
+        paths: node_indices
+            .into_par_iter()
+            .map(|x| {
+                let mut paths: DictMap<NodeIndex, Vec<NodeIndex>> =
+                    DictMap::with_capacity(graph.node_count());
+                let path_lenghts: Option<Vec<Option<f64>>> =
+                    bellman_ford(graph, x, |e| edge_cost(e.id()), Some(&mut paths)).unwrap();
```
Comment (mtreinish) [inline src/shortest_path/all_pairs_bellman_ford.rs:159]: Same comment as above

```suggestion
                if negative_cycle.read().unwrap() {
                    return (
                        x.index(),
                        PathMapping {
                            paths: DictMap::new(),
                        },
                    );
                }
                let mut paths: DictMap<NodeIndex, Vec<NodeIndex>> =
                    DictMap::with_capacity(graph.node_count());
                let path_lenghts: Option<Vec<Option<f64>>> =
                    bellman_ford(graph, x, |e| edge_cost(e.id()), Some(&mut paths)).unwrap();
```

File: src/shortest_path/all_pairs_bellman_ford.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+                if path_lenghts.is_none() {
```
Comment (mtreinish) [inline src/shortest_path/all_pairs_bellman_ford.rs:161]:
```suggestion
                if path_lengths.is_none() {
```

**Review by mtreinish — COMMENTED** [review]
Overall LGTM, this is pretty straighforward just mirroring the parallel `all_pair_*` shortest path algorithms we have for dijkstra and floyd warshall already. Just a couple small comments inline about improving the return time for negative cycles and a variable name typo.

File: src/shortest_path/all_pairs_bellman_ford.rs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                            opt_cost.map(|cost| (index, cost))
+                        } else {
+                            None
+                        }
+                    })
+                    .collect(),
+            };
+            (x.index(), out_map)
+        })
+        .collect();
```
[the remark this answers is not in the document]
Comment (IvanIsCoding) [reply src/shortest_path/all_pairs_bellman_ford.rs]: Done

File: src/shortest_path/all_pairs_bellman_ford.rs (hunk 3 above)
Comment (IvanIsCoding) [reply src/shortest_path/all_pairs_bellman_ford.rs]: I consistently mispell that word, thanks for catching it

File: src/shortest_path/all_pairs_bellman_ford.rs (hunk 2 above)
Comment (IvanIsCoding) [reply src/shortest_path/all_pairs_bellman_ford.rs]: Done

**Review by mtreinish — APPROVED** [review]
LGTM now, thanks for the quick update.

**Pull request merged**